azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,441 @@
+                logger.error(traceback.format_exc())
+
+    def _call_complete(self, event_data):
+        logger.info("Complete register from polling machine")
+        registration_result = event_data.args[0]
+
+        if self.on_registration_complete:
+            try:
+                self.on_registration_complete(registration_result)
+            except:  # noqa: E722 do not use bare 'except'
+                logger.error("Unexpected error calling on_registration_complete")
+                logger.error(traceback.format_exc())
+
+        callback = self._register_callback
+
+        if callback:
+            self._register_callback = None
+            try:
+                callback(registration_result, None)
+            except:  # noqa: E722 do not use bare 'except'
+                logger.error("Unexpected error calling callback supplied to register")
+                logger.error(traceback.format_exc())
+        # TODO : Should disconnect be by default called here ?
+
+    def _set_query_timer(self):
+        def time_up_query():
+            logger.error("Time is up for query timer")
+            self._query_timer.cancel()
+            # TimeoutError no defined in python 2
+            self._trig_error(ValueError("Time is up for query timer"))
+
+        self._query_timer = Timer(constant.DEFAULT_TIMEOUT_INTERVAL, time_up_query)
+        self._query_timer.start()
+
+    def _wait_for_interval(self, event_data):
+        def time_up_polling():
+            self._polling_timer.cancel()
+            logger.info("Done waiting for polling interval of {} secs".format(polling_interval))
+            if result.operation_id is None:
+                self._trig_send_register_request(event_data)
+            else:
+                self._trig_poll(event_data)
+
+        result = event_data.args[0]
+        polling_interval = (
+            constant.DEFAULT_POLLING_INTERVAL
+            if result.retry_after is None
+            else int(result.retry_after, 10)
+        )
+
+        self._polling_timer = Timer(polling_interval, time_up_polling)
+        logger.info("Waiting for " + str(constant.DEFAULT_POLLING_INTERVAL) + " secs")
+        self._polling_timer.start()  # This is waiting for that polling interval
+
+    def _decode_complete_json_response(self, query_result, response):
+        """
+        Decodes the complete json response for details regarding the registration process.
+        :param query_result: The partially formed result.
+        :param response: The complete response from the service
+        """
+        decoded_result = json.loads(response)
+
+        decoded_state = (
+            None
+            if "registrationState" not in decoded_result
+            else decoded_result["registrationState"]
+        )
+        registration_state = None
+        if decoded_state is not None:
+            # Everything needs to be converted to string explicitly for python 2
+            # as everything is by default a unicode character
```

Comment:
Not an issue, just a comment - we should do a teamwide sync and formalize on how we're handling this issue. So far we've just been doing it by ear, and there's probably some real inconsistency. We have no requirements defined for how this all should be handled.